clang/test/SemaCXX/cheri/cheri-capability-qualifier-dependent.cpp

```diff
@@ -0,0 +1,25 @@
+// RUN: %cheri_cc1 -o - %s -fsyntax-only -verify
+
+template<typename T>
+T __capability test_template(T p) {
+  // expected-note@-1 {{candidate template ignored: substitution failure [with T = int]: __capability only applies to pointers; type here is 'int'}}
+  // expected-note@-2 {{candidate template ignored: substitution failure [with T = int]: __capability only applies to pointers; type here is 'int'}}
+  T __capability c = (__cheri_tocap T __capability)p;
+  return c;
+}
+
+template void * __capability test_template<void *>(void *);
+template int __capability test_template<int>(int);
+// expected-error@-1 {{__capability only applies to pointers; type here is 'int'}}
+// expected-error@-2 {{explicit instantiation of 'test_template' does not refer to a function template, variable template, member function, member class, or static data member}}
+template int test_template<int>(int);
+// expected-error@-1 {{explicit instantiation of 'test_template' does not refer to a function template, variable template, member function, member class, or static data member}}
+
+template<typename T, typename E>
+E test_template2(T p) {
+  T __capability c = (__cheri_tocap T __capability)p;
+  return *c;
+}
+
```

Comment:
Maybe add another test case that isn't an explicit specialization?
```suggestion

void test_call() {
  (void) test_template(1.0f)
}
```

Reply:
It might also be a good idea to document what happens with types that already have __capability. Maybe something like:
```
void *__capability test_already_has_cap(void *__capability x) {
  return test_template(x)
}
```
My initial expectation would be that this works fine, but I guess a redundant qualifier warning might also be a valid option.